### PR TITLE
Add support for UEFI

### PR DESCRIPTION
**Pull request by tsoutsman**
Mostly changes to the build.

The only substantial addition to Theseus code is `nano_core/src/ uefi.rs`.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

### CODE REVIEW

File: kernel/nano_core/src/uefi.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    config.mappings.page_table_recursive =
+        Some(Mapping::FixedAddress(0o177777_776_000_000_000_0000));
+    config.kernel_stack_size = STACK_SIZE as u64;
+    config.serialize()
+};
+
+#[naked]
+#[no_mangle]
+#[link_section = ".init.text"]
+pub unsafe extern "C" fn _start(boot_info: &'static mut bootloader_api::BootInfo) {
```
Comment (kevinaboos) [inline kernel/nano_core/src/uefi.rs]: any particular reason why this is static mut? 
Also, how is it a rust-level reference and not a raw pointer if this is the first time we're entering Rust code? I understand they are interchangeable but it still conveys a misleading calling context.

File: kernel/nano_core/src/uefi.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        // page.
+        "sub rsp, 4096",
+        "call rust_entry",
+        "jmp KEXIT",
+        options(noreturn),
+    );
+}
+
+#[no_mangle]
+unsafe extern "C" fn rust_entry(
```
Comment (kevinaboos) [inline kernel/nano_core/src/uefi.rs]: please separate this into an unsafe function (which will deref the boot into pointer) and a safe function (called by the unsafe one) that calls the `nano_core` fn

File: kernel/nano_core/src/uefi.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    try_exit!(nano_core(
+        boot_info as &'static bootloader_api::BootInfo,
+        kernel_stack_start,
+    ));
+}
+
+#[naked]
+#[no_mangle]
+#[link_section = ".init.text"]
+pub unsafe extern "C" fn _error() {
```
Comment (kevinaboos) [inline kernel/nano_core/src/uefi.rs]: technically this function is not unsafe, only the interior asm block is.

**Review by kevinaboos — CHANGES_REQUESTED** [review]
looks good, left a few small comments

File: kernel/nano_core/src/uefi.rs (hunk 1 above)
Comment (tsoutsman) [reply kernel/nano_core/src/uefi.rs]: > any particular reason why this is static mut?

Changed

> Also, how is it a rust-level reference and not a raw pointer if this is the first time we're entering Rust code?

The bootloader runs rust code and directly calls `_start` with a Rust reference.

From <https://github.com/rust-osdev/bootloader>:
> Define an entry point function with the signature `fn kernel_main(boot_info: &'static mut bootloader_api::BootInfo) -> !`

File: kernel/nano_core/src/uefi.rs (hunk 2 above)
Comment (tsoutsman) [reply kernel/nano_core/src/uefi.rs]: As I mentioned above, we don't need to use raw pointers. The function is still unsafe, though as `double_fault_stack` has to be pointing to the end of a mapped page.

I've added the `unsafe_op_in_unsafe_fn` lint to the `nano_core` crate, which means that the unsafe operations in an unsafe function are clearly marked. Is the lint good enough?

File: kernel/nano_core/src/uefi.rs (hunk 1 above)
Comment (kevinaboos) [reply kernel/nano_core/src/uefi.rs]: ok i see, didn't realize we made the transition from Rust -> asm -> Rust again. Works for me.

**Review by kevinaboos — APPROVED** [review]

**Pull request merged**